src/shims/native_lib/mod.rs

```diff
@@ -276,6 +279,115 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
 
         interp_ok(())
     }
+
+    /// Extract the value from the result of reading a scalar or mplace from the machine,
+    /// and convert it to a `CArg`.
+    fn op_to_carg(&self, v: &OpTy<'tcx>) -> InterpResult<'tcx, CArg> {
+        let this = self.eval_context_ref();
+        let scalar = |v| interp_ok(this.read_immediate(v)?.to_scalar());
+        interp_ok(match v.layout.ty.kind() {
+            // If the primitive provided can be converted to a type matching the type pattern
+            // then create a `CArg` of this primitive value with the corresponding `CArg` constructor.
+            // the ints
+            ty::Int(IntTy::I8) => CPrimitive::Int8(scalar(v)?.to_i8()?).into(),
+            ty::Int(IntTy::I16) => CPrimitive::Int16(scalar(v)?.to_i16()?).into(),
+            ty::Int(IntTy::I32) => CPrimitive::Int32(scalar(v)?.to_i32()?).into(),
+            ty::Int(IntTy::I64) => CPrimitive::Int64(scalar(v)?.to_i64()?).into(),
+            ty::Int(IntTy::Isize) =>
+                CPrimitive::ISize(scalar(v)?.to_target_isize(this)?.try_into().unwrap()).into(),
+            // the uints
+            ty::Uint(UintTy::U8) => CPrimitive::UInt8(scalar(v)?.to_u8()?).into(),
+            ty::Uint(UintTy::U16) => CPrimitive::UInt16(scalar(v)?.to_u16()?).into(),
+            ty::Uint(UintTy::U32) => CPrimitive::UInt32(scalar(v)?.to_u32()?).into(),
+            ty::Uint(UintTy::U64) => CPrimitive::UInt64(scalar(v)?.to_u64()?).into(),
+            ty::Uint(UintTy::Usize) =>
+                CPrimitive::USize(scalar(v)?.to_target_usize(this)?.try_into().unwrap()).into(),
+            ty::RawPtr(..) => {
+                let s = scalar(v)?.to_pointer(this)?.addr();
+                // This relies on the `expose_provenance` in the `visit_reachable_allocs` callback
+                // above.
+                CPrimitive::RawPtr(std::ptr::with_exposed_provenance_mut(s.bytes_usize())).into()
+            }
+            // For ADTs, create a libffi::middle::Type from their fields.
+            ty::Adt(adt_def, args) => {
+                let strukt = this.adt_to_carg(v.layout.ty, *adt_def, args)?;
+
+                // The raw bytes backing this arg.
+                let bytes = match v.as_mplace_or_imm() {
```

Comment:
Can a struct ever be an imm? From my testing it seems like it's always an mplace here but I'm not sure